Support Flash iAP and Bootloader for GR-PEACH and GR-LYCHEE

TomoYamanaka opened the pull request:
### Description
Renesas added flash iap and bootloader feature to GR-PEACH and GR-LYCHEE.

Flash iAP was successful by `tests-mbed_drivers-flashiap` and `tests-mbed_hal-flash`.
Also Those was tested with https://github.com/ARMmbed/mbed-os-example-bootloader and was successful.

### Pull request type


    [ ] Fix
    [ ] Refactor
    [x] New target
    [ ] Feature
    [ ] Breaking change

0xc0170 commented:
/morph build

0xc0170 commented:
@TomoYamanaka Please review the build failures

TomoYamanaka committed:
Support Flash iAP for GR-PEACH and GR-LYCHEE

The mainly changes is below:
- Add flash_api.c
- Add the definition of SPI multi I/O Bus controller that is used for flash access
- Add "FLASH" as device feature
- Add the macro regarding information of the incorporated Flash
- Add the processing to expand code to RAM

TomoYamanaka commented:
I rebased my commits due to address the build failure.

cmonr reviewed (dismissed):
👍

cmonr commented:
/morph build

0xc0170 commented:
Please review exporter failure. There are multiple warnings for redefinition (that would be nice to be fixed), but more importantly here - 3 failures reported for ARMCC

TomoYamanaka commented:
Exporter occurs failure only IAR according the above report, so I modified the IAR definition for export.

TomoYamanaka commented:
With respect to the error message of Exporter for UVISION,
```
assembling irq_ca.S...
irq_ca.S: Error:  #5: cannot open source input file "__UVISION_VERSION SETA 521": No such file or directory
irq_ca.S: __UVISION_VERSION SETA 521: 0 warnings, 1 errorirq_ca.S: Error: A1067E: Output file specified as 'mbed-os\rtos\TARGET_CORTEX\rtx5\RTX\Source\TOOLCHAIN_ARM\TARGET_CORTEX_A\irq_ca.S', but it has already been specified as '.\build\irq_ca.o'
irq_ca.S: error: A1905U: Pre-processor step failed for '__UVISION_VERSION SETA 521'

```
I don' know why the message is outputed. 
The specified file is out of this time's changes renge, and I cannot find "__UVISION_VERSION SETA 521" file into mbed-os repo.
` mbed-os\rtos\TARGET_CORTEX\rtx5\RTX\Source\TOOLCHAIN_ARM\TARGET_CORTEX_A\irq_ca.S
`

Previously, although I updated this file for RTX5 update by cherry-picking from CMSIS_5 repo, I did not change it since then.
this error did not outputed in the PR of its updating.
https://github.com/ARMmbed/mbed-os/pull/5628#issuecomment-353559760

@toyowata Please help me to go ahead with the above.

cmonr commented:
@TomoYamanaka @toyowata On your testing machines, I would recommend rewinding this PR, and running the export build command, applying one commit from this PR at a time.

TomoYamanaka commented:
@cmonr
Could you tell me how to running the export build command for uvision's failure?

cmonr commented:
@TomoYamanaka Sure thing. It took longer than expected since I didn't want to give instructions that cause you to download all examples.

```
# Make examples directory, so that the export script knows where to populate project files
mkdir examples

# Clone example project
cd examples/
git clone https://github.com/ARMmbed/mbed-os-example-blinky

# Run the export script, -while- being in the examples directory
python -u ../mbed-os/tools/test/examples/examples.py --example mbed-os-example-blinky export IAR -m GR_LYCHEE
```

In windows, using the git bash shell should make all of these commands work as expected.
Let me know if that doesn't quite do the trick.

TomoYamanaka commented:
@cmonr 
Although I strive to reproduce the exporter failure by taking the advice from you, unfortunately, I can not create the environment correctly, and face the following message.

```
$ python -u ../mbed-os/tools/test/examples/examples.py --example mbed-os-example-blinky export UVISION -m GR_LYCHEE

Exporting example repos....

Exporting mbed-os-example-blinky GR_LYCHEE uvision
Scan: .
SUCCESS exporting mbed-os-example-blinky GR_LYCHEE uvision
Building mbed-os-example-blinky GR_LYCHEE uvision
Traceback (most recent call last):
  File "../mbed-os/tools/test/examples/examples.py", line 133, in <module>
    sys.exit(main())
  File "../mbed-os/tools/test/examples/examples.py", line 84, in main
    return args.fn(args, config, examples)
  File "../mbed-os/tools/test/examples/examples.py", line 90, in do_export
    results = lib.export_repos(config, args.ide, args.mcu, examples)
  File "C:\Work\Mission\mbed\GR-LYCHEE\src\Bootloader\exporter_chk2\mbed-os\tools\test\examples\examples_lib.py", line 296, in export_repos
    if EXPORTERS[ide].build(example_project_name, cleanup=False):
  File "C:\Work\Mission\mbed\GR-LYCHEE\src\Bootloader\exporter_chk2\mbed-os\tools\export\uvision\__init__.py", line 258, in build
    p = Popen(cmd, stdout=PIPE, stderr=PIPE)
  File "C:\Python27\lib\subprocess.py", line 711, in __init__
    errread, errwrite)
  File "C:\Python27\lib\subprocess.py", line 959, in _execute_child
    startupinfo)
WindowsError: [Error 2] No such file or directory
```


The procedure I tried is below. 

```
1. Creating working directory (same level as mbed-os remote repo)
$ mkdir examples
$ ls
exaples/  mbed-os/

2. Clone example project
$ cd exaples
$ git clone https://github.com/ARMmbed/mbed-os-example-blinky
$ cd mbed-os-example-blinky
$ mbed deploy

3. Run the export script
$ cd ..
$ ls
mbed-os-example-blinky/
$ python -u ../mbed-os/tools/test/examples/examples.py --example mbed-os-example-blinky export UVISION -m GR_LYCHEE
```

Could you give me your opinion about the above?

ghost commented:
Removing RC1 tag.  This is Target support, not new feature.  OK to go for Patch release.

TomoYamanaka commented:
@cmonr 
I appreciate if you can give me something comments this soon.

0xc0170 commented:
@TomoYamanaka it's been some time the build was run. If you can rebase, I would like to retrigger the build

TomoYamanaka committed:
Support Bootloader for GR-PEACH and GR-LYCHEE

The mainly changes is below:
- Update scatter file, linker file for bootloader support
- Update the file for RZ/A1 serial flash boot loader
- Add "device name" and "bootloader_supported" in targets.json

TomoYamanaka committed:
Update arm pack manager for GR-LYCHEE

I added the information of GR-LYCHEE for supporting bootloader in index.json file of arm pack manager.

TomoYamanaka committed:
Add the process for Cortex-A in mbed_applicaion

I added the process for Cortex-A in mbed_application.c because this process was for only Cortex-M.
Also I enabled the macro of MBED_APPLICATION_SUPPORT for Cortex-A.

TomoYamanaka committed:
Fix definitions of IAR exporter for GR-PEACH and GR-LYCHEE

TomoYamanaka commented:
@0xc0170  I rebased.

cmonr commented:
@TomoYamanaka Apologies for not responding yesterday, but it was a holiday in the states.

With the export command, could you attempt to build against a different IDE and/or target?

Since the project appears to export correctly, but not build, have you checked to see if project files exist within the blinky directory? If they do, are you able to open the project and build it in the IDE?

TomoYamanaka commented:
@cmonr 

> With the export command, could you attempt to build against a different IDE and/or target?

Although I tried the following commands, unfortunately, the python's error message remain to display.

```
$ python -u ../mbed-os/tools/test/examples/examples.py --example mbed-os-example-blinky export UVISION -m GR_LYCHEE
$ python -u ../mbed-os/tools/test/examples/examples.py --example mbed-os-example-blinky export UVISION -m RZ_A1H
$ python -u ../mbed-os/tools/test/examples/examples.py --example mbed-os-example-blinky export UVISION -m K64F

$ python -u ../mbed-os/tools/test/examples/examples.py --example mbed-os-example-blinky export IAR -m GR_LYCHEE
$ python -u ../mbed-os/tools/test/examples/examples.py --example mbed-os-example-blinky export IAR -m RZ_A1H
$ python -u ../mbed-os/tools/test/examples/examples.py --example mbed-os-example-blinky export IAR -m K64F
```

Is there any mistake in the above command?

cmonr commented:
@studavekar Can I get your help here?

@TomoYamanaka The commands appear correct. I'd like you to verify that both the uvision and iar tools are in your PATH environment variable.
For reference, my path variables for these tools are `C:\Keil_v5\UV4` and `C:\Program Files (x86)\IAR Systems\Embedded Workbench 7.5\common\bin`, respectively. The tool needs to be able to find the UV4.exe and IarBuild.exe executables.

studavekar commented:
> @studavekar Can I get your help here?
> 
> @TomoYamanaka The commands appear correct. I'd like you to verify that both the uvision and iar tools are in your PATH environment variable.
> For reference, my path variables for these tools are C:\Keil_v5\UV4 and C:\Program Files (x86)\IAR Systems\Embedded Workbench 7.5\common\bin, respectively. The tool needs to be able to find the UV4.exe and IarBuild.exe executables.

Looks like path variable should do the trick.

TomoYamanaka commented:
@cmonr 
Thank you for comments.

While I will check the above, I would like to confirm the message of exporter fail of UVISION.
```
assembling irq_ca.S...
irq_ca.S: Error:  #5: cannot open source input file "__UVISION_VERSION SETA 521": No such file or directory
irq_ca.S: __UVISION_VERSION SETA 521: 0 warnings, 1 errorirq_ca.S: Error: A1067E: Output file specified as 'mbed-os\rtos\TARGET_CORTEX\rtx5\RTX\Source\TOOLCHAIN_ARM\TARGET_CORTEX_A\irq_ca.S', but it has already been specified as '.\build\irq_ca.o'
irq_ca.S: error: A1905U: Pre-processor step failed for '__UVISION_VERSION SETA 521'

```

https://github.com/ARMmbed/mbed-os/pull/5826#issuecomment-360842610
According to the above PR#5826 comments, the exporter failure message of UVISION occured in the same way, it seems that it addressed by updating mbed-ci.

Could you give me anything comments?

toyowata commented:
@TomoYamanaka @0xc0170 @cmonr @studavekar 

I added the ```C:\Keil_v5\UV4``` to the ```PATH``` system environment variable and confirmed that no python runtime error (ralated to process creation).

Now, I got build error which is same as CI test log below:


```
Exporting example repos....

Exporting mbed-os-example-blinky RZ_A1H uvision
Scan: .
Warning: Multiple linker scripts detected: .\.link_script.sct -> .\mbed-os\targets\TARGET_RENESAS\TARGET_RZ_A1XX\TARGET_RZ_A1H\device\TOOLCHAIN_ARM_STD\MBRZA1H.sct
Scan: FEATURE_LWIP
SUCCESS exporting mbed-os-example-blinky RZ_A1H uvision
Building mbed-os-example-blinky RZ_A1H uvision
*** Using Compiler 'V5.06 update 6 (build 750)', folder: 'C:\Keil_v5\ARM\ARMCC\Bin'
Rebuild target 'mbed-os-example-blinky'
compiling aesni.c...
compiling arc4.c...

(snip)

assembling irq_ca.S...
irq_ca.S: Error:  #5: cannot open source input file "__UVISION_VERSION SETA 525": No such file or directory
irq_ca.S: __UVISION_VERSION SETA 525: 0 warnings, 1 errorirq_ca.S: Error: A1067E: Output file specified as 'mbed-os\rtos\TARGET_CORTEX\rtx5\RTX\Source\TOOLCHAIN_ARM\TARGET_CORTEX_A\irq_ca.S', but it has already been specified as '.\build\irq_ca.o'
irq_ca.S: error: A1905U: Pre-processor step failed for '__UVISION_VERSION SETA 525'
```

Then, I open uvision project file to build the project from the IDE (I also got same error).  In the Asm option panel, I found command line option below:

![export_1](https://user-images.githubusercontent.com/4926755/40775566-5bd174d4-6503-11e8-8d33-43c31597ee57.png)

```
[Misc Controls]
--cpreproc --cpreproc_opts=-D__ASSERT_MSG,-D__FPU_PRESENT,-DARM_MATH_CA9,-D__MBED_CMSIS_RTOS_CA9,-D__CMSIS_RTOS,-D__CORTEX_A9,-D__EVAL

[Assembler control string]
--cpu Cortex-A9 -g --apcs=interwork --cpreproc --cpreproc_opts=-D__ASSERT_MSG,-D__FPU_PRESENT,-DARM_MATH_CA9,-D__MBED_CMSIS_RTOS_CA9,-D__CMSIS_RTOS,-D__CORTEX_A9,-D__EVAL 
-I C:\Keil_v5\ARM\RV31\INC 
-I C:\Keil_v5\ARM\CMSIS\Include 
-I 
--pd "__UVISION_VERSION SETA 525" --list ".\BUILD\*.lst" --xref -o "*.o" --depend "*.d" 
```

According to armasm document, the ```--pd``` option is for predefined symbol, but it seems to be recognized as an input file (based on the error message).

http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0802b/BGBDAEGB.html

The problem seems to be a ```-I``` option with blank path  just above the ```--pd``` option line.

The ```Assembler control string``` in the Asm option dialog is not editable, so I cannot remove it.  However, when I checked ```No Auto Includes``` option in the same dialog, ```-I``` options are removed and I have compiled the blinky code successfully.

![export_2](https://user-images.githubusercontent.com/4926755/40775674-987a1800-6503-11e8-9d53-a7d45fc1065c.png)

```
*** Using Compiler 'V5.06 update 6 (build 750)', folder: 'C:\Keil_v5\ARM\ARMCC\Bin'
Build target 'mbed-os-example-blinky'
assembling irq_ca.S...
mbed-os\rtos\TARGET_CORTEX\rtx5\RTX\Source\TOOLCHAIN_ARM\TARGET_CORTEX_A\irq_ca.S(90): warning: A1786W: This instruction using SP is deprecated and may not work in future architecture revisions
mbed-os\rtos\TARGET_CORTEX\rtx5\RTX\Source\TOOLCHAIN_ARM\TARGET_CORTEX_A\irq_ca.S(124): warning: A1786W: This instruction using SP is deprecated and may not work in future architecture revisions
mbed-os\rtos\TARGET_CORTEX\rtx5\RTX\Source\TOOLCHAIN_ARM\TARGET_CORTEX_A\irq_ca.S(153): warning: A1786W: This instruction using SP is deprecated and may not work in future architecture revisions
mbed-os\rtos\TARGET_CORTEX\rtx5\RTX\Source\TOOLCHAIN_ARM\TARGET_CORTEX_A\irq_ca.S(296): warning: A1786W: This instruction using SP is deprecated and may not work in future architecture revisions
mbed-os\rtos\TARGET_CORTEX\rtx5\RTX\Source\TOOLCHAIN_ARM\TARGET_CORTEX_A\irq_ca.S(365): warning: A1786W: This instruction using SP is deprecated and may not work in future architecture revisions
mbed-os\rtos\TARGET_CORTEX\rtx5\RTX\Source\TOOLCHAIN_ARM\TARGET_CORTEX_A\irq_ca.S(428): warning: A1786W: This instruction using SP is deprecated and may not work in future architecture revisions
linking...
.\.link_script.sct(114): warning: L6314W: No section matches pattern *(HEAP).
.\.link_script.sct(124): warning: L6314W: No section matches pattern *(NC_DATA).
.\.link_script.sct(127): warning: L6329W: Pattern *(NC_BSS) only matches removed unused sections.
Program Size: Code=55140 RO-data=17796 RW-data=2652 ZI-data=122204  
Finished: 0 information, 3 warning and 0 error messages.
".\BUILD\mbed-os-example-blinky.axf" - 0 Error(s), 9 Warning(s).
Build Time Elapsed:  00:00:04
```

I think the error by CI test is not related this PR change, but someing in the exporter itself.

cmonr commented:
@toyowata A thought. The error looks like it might be something we've seen before. Do you know if you have the device pack installed for your devices?

@studavekar This looks familiar to https://github.com/ARMmbed/mbed-os/pull/5826#issuecomment-361419826. Thoughts?

toyowata commented:
@cmonr 

I have not installed particular pack file for this.  When I open the project file by uvision IDE, I got following error in the build output window.

```
*** Error - Project: 'mbed-os-example-blinky', Target: 'mbed-os-example-blinky' - 
*** Device: 'ARMCA9', Vendor: 'ARM', Dcore: 'Cortex-A9'
*** Cause - Dcore 'Cortex-A9' is not supported causing invalid Compiler/Debug/Flash setup.
*** Error - Project: 'mbed-os-example-blinky', Target: 'mbed-os-example-blinky' - 
*** Device: 'ARMCA9', Vendor: 'ARM', Dcore: 'Cortex-A9'
*** Cause - Dcore 'Cortex-A9' is not supported causing invalid Compiler/Debug/Flash setup.
```

I installed a pack file from [Renesas] - [RZ_A] - [Device Specific] - [Keil::RZ_DFP], but same error above.

0xc0170 commented:
@ARMmbed/mbed-os-tools Can you review https://github.com/ARMmbed/mbed-os/pull/6925#issuecomment-393479393 comment?

theotherjimmy commented on targets/TARGET_RENESAS/TARGET_RZ_A1XX/TARGET_GR_LYCHEE/device/TOOLCHAIN_IAR/MBRZA1LU.icf:
```diff
@@ -2,10 +2,20 @@
-define symbol __ICFEDIT_region_ROM_start__   = 0x18000000;
-define symbol __ICFEDIT_region_ROM_end__     = 0x187FFFFF;
+define symbol __ICFEDIT_region_ROM_start__   = MBED_APP_START;
+define symbol __ICFEDIT_region_ROM_end__     = MBED_APP_SIZE - 1;
```
Is this the end address? if so, It's calculated incorrectly.

theotherjimmy commented on targets/TARGET_RENESAS/TARGET_RZ_A1XX/TARGET_RZ_A1H/device/TOOLCHAIN_IAR/MBRZA1H.icf:
```diff
@@ -2,10 +2,20 @@
-define symbol __ICFEDIT_region_ROM_start__   = 0x18000000;
-define symbol __ICFEDIT_region_ROM_end__     = 0x187FFFFF;
+define symbol __ICFEDIT_region_ROM_start__   = MBED_APP_START;
+define symbol __ICFEDIT_region_ROM_end__     = MBED_APP_SIZE - 1;
```
Same comment + Why not take the opportunity to remove the duplicated linker scripts?

theotherjimmy reviewed (dismissed):
A few questions.

TomoYamanaka commented:
@theotherjimmy 

> ...ENESAS/TARGET_RZ_A1XX/TARGET_GR_LYCHEE/device/TOOLCHAIN_IAR/MBRZA1LU.icf
> Is this the end address? if so, It's calculated incorrectly.

I rebased because modified my commit based on your comments.

<br>

> ...ET_RENESAS/TARGET_RZ_A1XX/TARGET_RZ_A1H/device/TOOLCHAIN_IAR/MBRZA1H.icf
> Why not take the opportunity to remove the duplicated linker scripts?

As you see, there is hardly difference in those icf file, only the RAM size differs.
However, GR-PEACH and GR-LYCHEE incorporate different chips, so I would like to remain as is the duplicated linker scripts.

TomoYamanaka committed:
Update arm pack manager for GR-PEACH

I changed the information("sectors" and "memory") of GR-PEACH for supporting bootloader in index.json file of arm pack manager.

TomoYamanaka commented:
I added a commit to update arm pack manager for GR-PEACH.

TomoYamanaka commented:
I rebased to resolve a conflict of mbed_application.h, so could you review and re-trigger CI test?

TomoYamanaka commented:
@ARMmbed/mbed-os-tools 
> @ARMmbed/mbed-os-tools Can you review https://github.com/ARMmbed/mbed-os/pull/6925#issuecomment-393479393 comment?

Any update?

cmonr reviewed (dismissed):
Changes LGTM

cmonr commented:
/morph build

0xc0170 commented:
> https://github.com/ARMmbed/mbed-os/pull/6925#issuecomment-393479393

@ARMmbed/mbed-os-tools Can you please review this ? I am able to reproduce the issue (the latest exporter failed again). I installed latest device pack, and 3 errors in uvision.
@toyowata Thanks for all the details

Based on the changes here (adding new target device_has, thus new macro at least), are we hitting some limit in the uvision?

theotherjimmy reviewed (dismissed):
Changes look fine. I can't see any obvious reasons that the export build should be failing.

cmonr commented:
Looking into it now. Was able to reproduce the error on my system as well. Have asked for more information from those closer to the uVision product.

cmonr commented:
@ARMmbed/mbed-os-maintainers @TomoYamanaka  I think I know what's going on now.
After talking with @TacoGrandeTX, I was introduced to this page: http://www2.keil.com/mdk5/selector

That page indicates that ARM Cortex-A5/A7/A9 support only exists in DS-MDK IDE.


@theotherjimmy Is there a way to indicate that a target/board is not supported with an exporter?
@studavekar Is there a way within the export test to bypass a specific export test for a particular target/board that you know of?

theotherjimmy commented:
@cmonr Yes. That's the responsibility of the `is_target_supported` class-method on the exporter. It looks like we need to update that to filter on supported MCU arches.

0xc0170 commented:
> That page indicates that ARM Cortex-A5/A7/A9 support only exists in DS-MDK IDE.

That is what I  also initially thought, it explains the Dcore error shared above (also I noticed it locally). Then we shall fix.

> @cmonr Yes. That's the responsibility of the is_target_supported class-method on the exporter. It looks like we need to update that to filter on supported MCU arches.

👍 

@TomoYamanaka Once fix is in place, we restart CI and should be all green.

TomoYamanaka commented:
I will appreciate if you can accelerate the resolution because this is a blocker for Renesas user's activity.

cmonr commented:
@TomoYamanaka It is high in my queue of work to get a PR in that will allow this to pass CI. Once that PR is made, I'll link it here.

cmonr commented:
@TomoYamanaka https://github.com/ARMmbed/mbed-os/pull/7295 has been created.

TomoYamanaka commented:
The exporter's problem was resolved, and I rebased to resolve a conflict of "tools/arm_pack_manager/index.json".
So could you review and re-trigger CI test again?

0xc0170 commented:
/morph build

cmonr commented:
#ghostinthemachine
